### PR TITLE
Do not mark overloaded functions as shadowing

### LINKED ISSUE #2676: Function overloads are reported as shadowing

### PR DESCRIPTION
**Pull request by axic**
Fixes #2676.

### CODE REVIEW

File: libsolidity/analysis/NameAndTypeResolver.cpp [hunk 1]
```diff
@@ -439,6 +439,23 @@ DeclarationRegistrationHelper::DeclarationRegistrationHelper(
...
+	Declaration const& _declaration2
+)
+{
+	auto const* function1 = dynamic_cast<FunctionDefinition const*>(&_declaration1);
+	auto const* function2 = dynamic_cast<FunctionDefinition const*>(&_declaration2);
+
+	if (!function1 || !function2)
+		return false;
+
+	if (function1->parameters() != function2->parameters())
```
Comment (chriseth) [inline libsolidity/analysis/NameAndTypeResolver.cpp]: I think in other places of the code, the function `hasEqualArgumentTypes` is used.

**Comment by chriseth** [comment]
I will find a test case that fails on my comment and then fix it.

File: libsolidity/analysis/NameAndTypeResolver.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/analysis/NameAndTypeResolver.cpp]: So they will actually always be different. The question is: Do we actually need to check this?

Comment (chriseth) [reply libsolidity/analysis/NameAndTypeResolver.cpp]: So the test should be some function that shadows e.g. the global sha3.

Comment (chriseth) [reply libsolidity/analysis/NameAndTypeResolver.cpp]: or perhaps ecrecover because it actually has a signatre.

Comment (axic) [reply libsolidity/analysis/NameAndTypeResolver.cpp]: If it is overloaded the types are the same?

Comment (axic) [inline libsolidity/analysis/NameAndTypeResolver.cpp]: Those aren't handled in this conditional, they are part of the "builtin" branch.

Comment (chriseth) [reply libsolidity/analysis/NameAndTypeResolver.cpp]: This comparison compares vectors of AST nodes. They will be different if the functions are different.

**Comment by chriseth** [comment]
I'm changing the condition to "warn about shadowing if the parent scope (or one of its parent scopes) contains an identifier of the same name". I think this should catch all cases and does not introduce any false positives. I will check if this is true for the global scope.

**Comment by axic** [comment]
Do we need to add more tests to this then?

**Comment by chriseth** [comment]
Yes, I think we should.

**Pull request merged**